dynamodb/README.md

```diff
@@ -0,0 +1,136 @@
+| `Map`              | `STRING`                   | json encoding                             |
+| `StringSet`        | `ARRAY<STRING>`            |                                           |
+| `NumberSet`        | `ARRAY<NUMERIC or STRING>` |                                           |
+| `BinarySet`        | `ARRAY<BYTES>`             |                                           |
+
+We discuss these, as well as other limits and notes on schema conversion, in the
+following sections.
+
+### Schema Inference
+
+DynamoDB is schemaless database: other than a primary index and optional
+secondary index, column names and types are essentially unconstrained
+and can vary from one row to the next.
+
+However, many customers use DynamoDB in a consistent, structured way
+with a fairly well defined set of columns and types. Our HarbourBridge support
+for DynamoDB focuses on this use-case, and we construct a Spanner schema
+by inspecting table data.
+
+For small tables, we inspect all rows of the table. For large tables, scanning
+the entire table would be extremely expensive and slow, and so we only inspect
+the first N rows (defined by the flag `schema-sample-size`) from the table scan.
+While DynamoDB doesn't return scan results in order, they might not be a truly
+random sample of rows. However, the alternative of randomly sampling rows
+would be much more expensive.
+
+Columns with consistent types are assigned Spanner types as detailed below.
+Columns without a consistent type are mapped to STRING.
+
+#### `Number`
+
+In most cases, we map the Number type in DynamoDB to Spanner's Numeric type.
+However, since the [range of Numeric](https://cloud.google.com/spanner/docs/storing-numeric-data)
+in Cloud Spanner is smaller than the [range of Number](https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/HowItWorks.NamingRulesDataTypes.html)
+in DynamoDB, this conversion could result in out of range with potential
+precision loss. To address this possibility, we try to convert the sample data,
+and if it consistently fails, we choose STRING type for the column.
+
+#### `Null` Data Type
+
+In DynamoDB, a column can have a Null data type that represents an unknown or
+undefined state. Also, each row defines its own schema for columns (not for
+primary keys). So columns can be absent in rows. 
+
+We treat the above two cases the same as a Null value in Cloud Spanner. The
+cases that a column contains a Null value or a column is not present is an
+indication that this column should be nullable.
+
+#### `List` and `Map`
+
+In Cloud Spanner, the most similar type to List and Map is
+[STRUCT](https://cloud.google.com/spanner/docs/data-types#struct_type), but it
+is not a valid column type (available for query but not for storage).
+Therefore, we encode them into a json string. 
+
+#### Occasional Errors
+
+As no schema is forced when writing to a DynamoDB table, it can happen that a
```

Comment:
Suggest rephrasing:

To prevent a few spurious rows from disrupting our schema construction
for an otherwise consistently typed table, we define an error threshold: if
the percentage of the data type is lower than or equal to an extremely
low value (0.1%), then we consider the data type as an occasional error.
Such rows are not considered as a candidate type for the column.